File: src/updater.ts
import { saveSlot } from "./progress";
import { useGame } from "./store";

// Desktop auto-update bridge.
//
// Three layers have to line up before any of this does something, and every
// one of them can be absent:
//
//   1. The app is running inside the Tauri shell at all (not the web build,
//      not the Capacitor shells).
//   2. That shell was compiled with the `updater` cargo feature.
//   3. It is not a Steam install — src-tauri/src/updater.rs refuses to
//      register the commands there, because a self-update would desynchronize
//      the SteamPipe depot.
//
// Only (1) is detectable from here, so (2) and (3) surface as "command not
// found" from `invoke` and are treated exactly like "no update": silent
// no-op. Nothing in this module ever throws at its caller.
//
// The Tauri API packages are behind dynamic imports so the web bundle never
// downloads them.

/** Payload of `updater_check` — mirrors `UpdateInfo` in src-tauri/src/updater.rs. */
export type UpdateInfo = {
  version: string;
  currentVersion: string;
  notes: string | null;
  date: string | null;
};

export type DownloadProgress = { downloaded: number; total: number | null };

/** Emitted by the Rust `on_before_exit` hook. Keep in sync with updater.rs. */
const BEFORE_EXIT_EVENT = "updater://before-exit";
const PROGRESS_EVENT = "updater://download-progress";

/**
 * Cheap synchronous probe, so the web build never even fetches the
 * `@tauri-apps/api` chunk. `globalThis.isTauri` is the flag Tauri's injected
 * IPC bootstrap sets, and is exactly what the package's own `isTauri()` reads.
 */
export const isTauriShell = (): boolean => {
  try {
    return !!(globalThis as { isTauri?: boolean }).isTauri;
  } catch {
    return false;
  }
};

const api = async () => {
  if (!isTauriShell()) return null;
  const [core, event] = await Promise.all([
    import("@tauri-apps/api/core"),
    import("@tauri-apps/api/event"),
  ]);
  return core.isTauri() ? { core, event } : null;
};

/**
 * Re-persists the active save slot.
 *
 * Progress writes are already synchronous and write-through (see
 * `persistProgress` in store.ts), so this is belt-and-braces rather than a
 * flush of buffered state — there is no buffer. What it buys is a fresh
 * `localStorage.setItem` immediately before the process dies, which is the
 * best any web-storage-backed save layer can do: there is no API to force the
 * WebView to fsync, so the window between the write and the disk is the
 * WebView's to decide.
 */
const persistNow = (): void => {
  try {
    const { activeSlot, progress } = useGame.getState();
    if (activeSlot !== null) saveSlot(activeSlot, progress);
  } catch {
    // Never let a save failure stall the installer.
  }
};

/**
 * Asks the update endpoint whether a newer release exists.
 *
 * Resolves to `null` for "already current" and for every unavailable-here
 * case; callers do not need to distinguish them.
 */
export const checkForUpdate = async (): Promise<UpdateInfo | null> => {
  try {
    const tauri = await api();
    if (!tauri) return null;
    return await tauri.core.invoke<UpdateInfo | null>("updater_check");
  } catch {
    return null;
  }
};

/**
 * Downloads and installs the update reported by the last `checkForUpdate`,
 * then restarts.
 *
 * Resolves only on failure — on success the process is replaced (macOS/Linux)
 * or force-exited by the installer (Windows). Returns `false` when the install
 * did not happen.
 */
export const installUpdate = async (
  onProgress?: (progress: DownloadProgress) => void,
): Promise<boolean> => {
  const tauri = await api();
  if (!tauri) return false;

  const unlisteners: Array<() => void> = [];
  try {
    // Windows installers force-exit the process partway through. This is the
    // only warning the frontend gets, and the Rust side blocks for a few
    // hundred milliseconds after emitting it — so the handler must be
    // synchronous. Registered per install rather than at boot: outside an
    // install there is nothing to hear.
    unlisteners.push(await tauri.event.listen(BEFORE_EXIT_EVENT, persistNow));
    if (onProgress) {
      unlisteners.push(
        await tauri.event.listen<DownloadProgress>(PROGRESS_EVENT, (e) => onProgress(e.payload)),
      );
    }

    // Save before the download even starts, so the guaranteed-synchronous path
    // covers the common case and the before-exit hook is only the backstop for
    // whatever the player did during the download.
    persistNow();

    await tauri.core.invoke("updater_install");
    return true;
  } catch {
    return false;
  } finally {
    for (const off of unlisteners) off();
  }
};

File: src-tauri/src/updater.rs
//! Tauri v2 auto-updater wiring.
//!
//! Compiled only when the `updater` Cargo feature is on (see Cargo.toml for
//! why that feature is opt-in) and only on desktop targets.
//!
//! ## Why the update flow is driven from Rust instead of the JS plugin API
//!
//! `tauri-plugin-updater`'s `download_and_install` IPC command builds its
//! updater through `UpdaterExt::updater_builder()`, which hardcodes
//! `on_before_exit(|| app.cleanup_before_exit())`. There is no way to add our
//! own hook to that path. On Windows the installer force-exits the process
//! (`std::process::exit(0)`) the moment it starts, so without our own hook the
//! frontend never learns the app is about to die and never gets to persist.
//!
//! So the plugin is still registered — it holds the `pubkey`/`endpoints`
//! config that `updater_builder()` reads — but the two commands the frontend
//! calls are ours. As a side effect they need no capability entries: Tauri v2
//! only ACL-checks `plugin:`-prefixed commands (and app commands when the app
//! ships its own `permissions/` manifest, which this one does not).

use std::sync::Mutex;
use std::time::Duration;

use serde::Serialize;
use tauri::{AppHandle, Emitter, Runtime, State};
use tauri_plugin_updater::{Update, UpdaterExt};

/// Emitted just before a Windows installer force-exits the process. The
/// frontend's listener should do synchronous persistence only.
pub const BEFORE_EXIT_EVENT: &str = "updater://before-exit";
/// Emitted while the update payload downloads.
pub const PROGRESS_EVENT: &str = "updater://download-progress";

/// How long we block the installer thread after emitting `BEFORE_EXIT_EVENT`.
///
/// This is a best-effort window, not a guarantee: `Emitter::emit` hands the
/// payload to the webview asynchronously, and the WebView's localStorage is
/// itself flushed to disk on its own schedule. Long enough to matter, short
/// enough that a wedged renderer does not visibly stall the install.
const BEFORE_EXIT_GRACE: Duration = Duration::from_millis(600);

/// The `Update` handle produced by a check, parked until the user accepts.
#[derive(Default)]
struct PendingUpdate(Mutex<Option<Update>>);

#[derive(Serialize, Clone)]
#[serde(rename_all = "camelCase")]
struct UpdateInfo {
    version: String,
    current_version: String,
    notes: Option<String>,
    date: Option<String>,
}

#[derive(Serialize, Clone)]
#[serde(rename_all = "camelCase")]
struct DownloadProgress {
    downloaded: u64,
    total: Option<u64>,
}

/// True when this process looks like it was launched from a Steam install.
///
/// Second layer behind the `updater` Cargo feature. It is load-bearing rather
/// than decorative here: `steam.yml` does not build anything, it reuses the
/// `*-portable` artifacts produced by the same `tauri build` runs that produce
/// the installers, so the depot binary and the installer binary are the same
/// binary. The compile-time feature cannot tell them apart; this can.
fn is_steam_runtime() -> bool {
    // Set by the Steam client for every process it launches.
    for key in [
        "SteamAppId",
        "SteamGameId",
        "SteamOverlayGameId",
        "SteamClientLaunch",
    ] {
        if std::env::var_os(key).is_some_and(|v| !v.is_empty()) {
            return true;
        }
    }

    let Ok(exe) = std::env::current_exe() else {
        return false;
    };

    // Dropped next to the executable so the Steamworks SDK can attach when the
    // game is started outside the client during development.
    if exe
        .parent()
        .is_some_and(|dir| dir.join("steam_appid.txt").exists())
    {
        return true;
    }

    // .../Steam/steamapps/common/<Game>/... on all three desktop platforms.
    exe.components()
        .any(|c| c.as_os_str().eq_ignore_ascii_case("steamapps"))
}

/// Asks the configured endpoint whether a newer version exists.
///
/// Returns `Ok(None)` when the app is already current. The resulting `Update`
/// is parked in app state so `updater_install` can act on the exact release
/// the user was shown.
#[tauri::command]
async fn updater_check<R: Runtime>(
    app: AppHandle<R>,
    pending: State<'_, PendingUpdate>,
) -> Result<Option<UpdateInfo>, String> {
    let handle = app.clone();
    let updater = app
        .updater_builder()
        // Overrides the plugin's default hook, which is exactly
        // `cleanup_before_exit()` — so we have to call that ourselves.
        .on_before_exit(move || {
            let _ = handle.emit(BEFORE_EXIT_EVENT, ());
            std::thread::sleep(BEFORE_EXIT_GRACE);
            handle.cleanup_before_exit();
        })
        .build()
        .map_err(|e| e.to_string())?;

    let update = updater.check().await.map_err(|e| e.to_string())?;

    let info = update.as_ref().map(|u| UpdateInfo {
        version: u.version.clone(),
        current_version: u.current_version.clone(),
        notes: u.body.clone(),
        date: u.date.map(|d| d.to_string()),
    });

    *pending.0.lock().map_err(|e| e.to_string())? = update;
    Ok(info)
}

/// Downloads and installs the update parked by the last `updater_check`.
#[tauri::command]
async fn updater_install<R: Runtime>(
    app: AppHandle<R>,
    pending: State<'_, PendingUpdate>,
) -> Result<(), String> {
    let update = {
        let mut slot = pending.0.lock().map_err(|e| e.to_string())?;
        slot.take()
    }
    .ok_or_else(|| "no pending update — run updater_check first".to_string())?;

    let progress_app = app.clone();
    let mut downloaded: u64 = 0;
    update
        .download_and_install(
            move |chunk, total| {
                downloaded += chunk as u64;
                let _ = progress_app.emit(PROGRESS_EVENT, DownloadProgress { downloaded, total });
            },
            || {},
        )
        .await
        .map_err(|e| e.to_string())?;

    // Windows never gets here: the installer force-exits the process through
    // the `on_before_exit` hook above. macOS and Linux swap the bundle in
    // place and return, so the relaunch is ours to do.
    app.restart();
}

/// Registers the plugin and the two commands, unless this looks like a Steam
/// install — in which case nothing is registered and every `invoke` from the
/// frontend fails with "command not found", which the frontend treats as
/// "updates are not available here".
pub fn register<R: Runtime>(builder: tauri::Builder<R>) -> tauri::Builder<R> {
    if is_steam_runtime() {
        eprintln!("mesozoic-protocol: Steam runtime detected — in-app updater disabled");
        return builder;
    }

    builder
        .plugin(tauri_plugin_updater::Builder::new().build())
        .manage(PendingUpdate::default())
        .invoke_handler(tauri::generate_handler![updater_check, updater_install])
}
